ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java

```diff
@@ -411,28 +448,132 @@ public void disconnectWorstUselessPeer() {
+                trailingPeerRequirementsSupplier.get();
+            if (trailingPeerRequirements != null) {
+              if (peer.chainState().getEstimatedHeight()
+                  < trailingPeerRequirements.getMinimumHeightToBeUpToDate()) {
+                if (!(getNumTrailingPeers(trailingPeerRequirements.getMinimumHeightToBeUpToDate())
+                    < trailingPeerRequirements.getMaxTrailingPeers())) {
+                  LOG.atTrace()
+                      .setMessage(
+                          "Adding trailing peer {} would exceed max trailing peers {}. Disconnecting...")
+                      .addArgument(peer.getLoggableId())
+                      .addArgument(trailingPeerRequirements.getMaxTrailingPeers())
+                      .log();
+                  peer.disconnect(
+                      DisconnectMessage.DisconnectReason.USELESS_PEER_EXCEEDS_TRAILING_PEERS);
+                  return;
+                }
+              }
+            }
+
+            peer.chainState().updateHeightEstimate(peerHeadBlockHeader.getNumber());
+            CompletableFuture<Void> isServingSnapFuture;
+            if (SyncMode.isCheckpointSync(syncMode) || SyncMode.isSnapSync(syncMode)) {
+              // even if we have finished the snap sync, we still want to know if the peer is a snap
+              // server
+              isServingSnapFuture =
+                  CompletableFuture.runAsync(
+                      () -> {
+                        try {
+                          checkIsSnapServer(peer, peerHeadBlockHeader);
+                        } catch (Exception e) {
+                          throw new RuntimeException(e);
+                        }
+                      });
+            } else {
+              isServingSnapFuture = CompletableFuture.completedFuture(null);
+            }
+            isServingSnapFuture.thenRun(
+                () -> {
+                  if (!peer.getConnection().isDisconnected() && addPeerToEthPeers(peer)) {
+                    connectedPeersCounter.inc();
+                    connectCallbacks.forEach(cb -> cb.onPeerConnected(peer));
+                  }
+                });
+          }
+        });
+  }
+
+  private void checkIsSnapServer(final EthPeer peer, final BlockHeader peersHeadBlockHeader) {
+    if (peer.getAgreedCapabilities().contains(SnapProtocol.SNAP1)) {
+      if (snapServerChecker != null) {
+        // set that peer is a snap server for doing the test
+        peer.setIsServingSnap(true);
+        Boolean isServer;
+        try {
+          isServer = snapServerChecker.check(peer, peersHeadBlockHeader).get(6L, TimeUnit.SECONDS);
+        } catch (Exception e) {
+          LOG.atTrace()
+              .setMessage("Error checking if peer {} is a snap server. Setting to false.")
+              .addArgument(peer.getLoggableId())
+              .log();
+          peer.setIsServingSnap(false);
+          return;
+        }
+        peer.setIsServingSnap(isServer);
```

Comment:
nit: L570 and L571 can be moved inside try block after L561.

Reply:
I prefer it this way, to make it a bit easier to understand.